src/main/java/com/nvidia/spark/rapids/jni/GpuTimeZoneDB.java

```diff
@@ -34,64 +34,78 @@
-    fixedTransitionsFuture = new CompletableFuture<>();
+  private GpuTimeZoneDB() {
   }
   
-  private static GpuTimeZoneDB instance = new GpuTimeZoneDB();
+  private static final GpuTimeZoneDB instance = new GpuTimeZoneDB();
   // This method is default visibility for testing purposes only. The instance will be never be exposed publicly
   // for this class.
   static GpuTimeZoneDB getInstance() {
     return instance;
   }
-  
+
+  /**
+   * This should be called on startup of an executor.
+   */
+  public static void cacheDatabaseAsync() {
+    Runnable runnable = () -> {
+      try {
+        cacheDatabase();
+      } catch (Exception e) {
+        // ignore
+      }
+    };
+    Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+    thread.setName("gpu-timezone-database-0");
+    thread.setDaemon(true);
+    thread.start();
+  }
+
   /**
-   * Start to cache the database. This should be called on startup of an executor. It should start
-   * to cache the data on the CPU in a background thread. It should return immediately and allow the
-   * other APIs to be called. Depending on what we want to do we can have the other APIs block
-   * until this is done caching, or we can have private APIs that would let us load and use specific
-   * parts of the database. I prefer the former solution at least until we see a performance hit
-   * where we are waiting on the database to finish loading.
+   * cache the database.
    */
   public static void cacheDatabase() {
-    synchronized (instance) {
-      if (!instance.isLoaded()) {
-        Executor executor = Executors.newSingleThreadExecutor(
-          new ThreadFactory() {
-            private ThreadFactory defaultFactory = Executors.defaultThreadFactory();
+    instance.cacheDatabaseImpl();
+  }
 
-            @Override
-            public Thread newThread(Runnable r) {
-              Thread thread = defaultFactory.newThread(r);
-              thread.setName("gpu-timezone-database-0");
-              thread.setDaemon(true);
-              return thread;
-            }
-          });
-        instance.loadData(executor);
+  public static void shutdown() {
+    instance.shutdownImpl();
+  }
+
+  private synchronized void cacheDatabaseImpl() {
+    if (cacheEmpty) {
+      try {
+        loadData();
+        cacheEmpty = false;
+      } catch (Exception e) {
+        // if it has exception, try to close the cache
+        try (HostColumnVector hcv = getHostFixedTransitions()) {
+          // automatically closed
+        }
+        zoneIdToTable.clear();
+        fixedTransitions = null;
+        zoneIdToTable = null;
+        throw e;
       }
     }
   }
 
-
-  public static void shutdown() {
-    if (instance.isLoaded()) {
-      instance.close();
-      // Recreate a new instance to reload the database if necessary
-      instance = new GpuTimeZoneDB();
+  public synchronized void shutdownImpl() {
+    if (!cacheEmpty) {
```

Comment:
Do we have a potential race condition here where a shutdown occurs very quickly after a startup, so quickly (or database load is so slow) that we don't think the cache is there but it will be once the async thread completes?  Seems like we need to also check if there's a pending async load and wait on it if that's the case.

Reply:
The caches APIs are synchronized, so multiple callings will be serial and wait other to be done.
```scala
private synchronized void cacheDatabaseImpl() {
private synchronized void shutdownImpl() {
```

Reply:
`synchronized` doesn't solve the race I'm talking about.  All `synchronized` does is guarantee two threads aren't in that method at the same time, but it does nothing to ensure what _order_ those threads call it.  If two or more threads are trying to do the same thing (e.g.; initialize the cache), we're fine, since we don't really care about the order of the threads since the operation they are performing is identical and idempotent.  However if the threads are doing _different_ operations, then we care very much about the order of operations.

If one thread is trying to start the cache and the other is trying to shutdown, we could leave the database in a cached state despite trying to shut it down.  Here's a simplified example:
- On startup, we launch the background thread to initialize the cache
- Before the background thread is able to boot up and call the cache init method, the shutdown method is called (e.g.: very quick test that doesn't involve time DB and tears down on test completion, error on startup, whatever)
- shutdown is called _first_, notices the cache is not initialized, does nothing.
- background thread finally gets around to calling cache init (note there is no guarantee how quick this happens) and proceeds to initialize the cache
- Now we're in a state where the application thinks the time DB cache has been shutdown and freed, but in reality it has not and we've "leaked" it in a sense.

One way to solve this is to set a boolean flag before we start the background thread.  The background caching thread would atomically clear this flag and call the cache init method when it runs.  The shutdown method would check for this flag being set, and if it is, release the lock and wait on the background thread to complete before proceeding to retry the shutdown.  We can optimize it a bit by having shutdown set a flag after the cache clear has occurred indicating a shutdown has happened.  The background init thread can check for this flag being set and skip cache init when that happens.

Reply:
A simpler approach is to have a boolean flag indicating a shutdown has occurred.  Caching the database will clear this flag.  The background thread avoids initializing the cache if the shutdown flag has been set.  The shutdown method always wait on the background init Thread (outside of the lock) if it's not null, setting it to null afterwards.

Reply:
The cache is a global singleton, and we do not want to shutdown it unless the Plugin is shutdown.
It’s almost not a scenario that one thread is caching cache and other is shutdown the cache concurrently.
 
Currently test cases do not call cacheDatabase/shutdown concurrently, see the following:
```scala
public class TimeZoneTest {
  @BeforeAll
  static void cacheTimezoneDatabase() {
    GpuTimeZoneDB.cacheDatabase();
  }
  
  @AfterAll
  static void cleanup() {
    GpuTimeZoneDB.shutdown();
  }
```
One potential corner case is:
Another time zone test like `TimeZoneTest2` doing the same like `TimeZoneTest`.
The following sequence will cause stale cache as you mentioned.
- TimeZoneTest.cache
- do TimeZoneTest.test1
- TimeZoneTest2.cache
- TimeZoneTest.shutdown
- do TimeZoneTest2.test2:  cache is shutdown, will cause a problem.
- TimeZoneTest2.shutdown

Typically production usage is:
Call `cacheDatabaseAsync` start the Plugin
Call `shutdown` when  stop the Plugin.

And another solution:
Remove the `shutdown` and mark the host memory column as [noWarnLeakExpected](https://github.com/rapidsai/cudf/blob/85acdc640701940e47b3969b14a811f33e7faf5b/java/src/main/java/ai/rapids/cudf/MemoryCleaner.java#L139), this will disable the cache leak warning for the global singleton host column.When the JVM exits the host memory column will be cleaned.

Do we need to do the implementation you mentioned for the rare use case?

Reply:
I'm not worried about the production use-case but rather the potential corner cases with tests that may shutdown and restart the database.  It's really not that hard to fix this race condition (I posted the details for two approaches above, happy to provide the patch for either), so why not close that race?  Even if we can prove to ourselves it's never going to be a race today, it doesn't guarantee someone won't change the way this is used to expose the race in the future.

Reply:
got it, I'll fix it.

Reply:
>One way to solve this is to set a boolean flag before we start the background thread. The background caching thread would atomically clear this flag and call the cache init method when it runs. The shutdown method would check for this flag being set, and if it is, release the lock and wait on the background thread to complete before proceeding to retry the shutdown. We can optimize it a bit by having shutdown set a flag after the cache clear has occurred indicating a shutdown has happened. The background init thread can check for this flag being set and skip cache init when that happens.

Now use this approach you mentioned.
Please review.
Changes are:
- Move all sync code into static methods.
- You mentioned changes